src/input_utils.py

```diff
@@ -193,21 +222,97 @@ def _decode_fn(record):
-    rest_len = max_seq_len - seq_len
-
-    # Create segment_ids
-    segment_ids = [tf.ones(num_image_wordpieces, dtype=tf.int32),
-                   tf.ones(num_text_wordpieces, dtype=tf.int32)*2,
-                   tf.zeros(rest_len, dtype=tf.int32)]
-    example['segment_ids'] = tf.concat(segment_ids, axis=0)
-
-    # Create input_mask
-    helper = tf.range(max_seq_len)
-    example['input_mask'] = tf.where(helper < seq_len, 1, 0)
+    example['num_image_wordpieces'] = 2 + num_patch_per_row**2
+    example['num_text_wordpieces'] = get_num_wordpieces(text_input_ids)
 
     return example
 
   return _decode_fn
+
+
+def make_relative_transformer_side_inputs(
+  long_breakpoints: tf.Tensor,
+  relative_pos_max_distance: int,
+  name: Optional[str] = None) -> RelativeTransformerSideInputs:
+  """Makes relative transformer side input tensors
+
+  Args:
+    long_breakpoints: <int32>[batch_size, long_seq_len] Tensor of ending
+      breakpoints separating different packed examples.
+    relative_pos_max_distance: Maximum distance to use for relative position
+      representations. All larger distances will be clipped to this value. Use
+      0 to skip relative position representations entirely.
+    name: A name for the operation (optional).
+
+  Returns:
+    A `RelativeTransformerSideInputs` with all relevant tensors set.
+  """
+
+  with tf.name_scope(name or 'make_relative_transformer_side_inputs'):
+    long_breakpoints = tf.convert_to_tensor(long_breakpoints)
+    long_example_ids = tf.cumsum(long_breakpoints, axis=-1, reverse=True)
+    long_example_ids = tf.convert_to_tensor(long_example_ids)
+    long_seq_len = tf_utils.get_shape_list(long_example_ids)[1]
+    att_mask = etc_feature_utils.make_segmented_att_mask(long_example_ids)
+    batch_size = tf.shape(long_example_ids)[0]
+
+    relative_att_ids = None
+    if relative_pos_max_distance > 0:
+      relative_pos_generator = etc_feature_utils.RelativePositionGenerator(
+          relative_pos_max_distance)
+      relative_att_ids = relative_pos_generator.make_relative_att_ids(
+          seq_len=long_seq_len,
+          batch_size=batch_size)
+
+    return RelativeTransformerSideInputs(
+        att_mask=att_mask,
+        relative_att_ids=relative_att_ids)
+
+
+def add_side_input_features(
+  input_config: PretrainInputConfig,
+  model_config: EtcConfig,
+  features: Mapping[str, tf.Tensor]) -> Mapping[str, tf.Tensor]:
+  """Replaces raw input features with derived ETC side inputs.
+
+  This function is meant to be called as part of a Dataset pipeline.
+
+  Args:
+    model_config: A EtcConfig.
+    features: A dictionary of Tensor features, crucially including
+      `long_breakpoints`, `num_image_wordpieces`, `num_text_wordpieces`.
+
+  Returns:
+    A new `features` dictionary with side inputs.
+  """
+
+  features = dict(features)
+
+  img_wp = features['num_image_wordpieces']
+  txt_wp = features['num_text_wordpieces']
+  seq_len = tf.math.add(img_wp, txt_wp)
```

Comment:
Nit: would this suffice?
```suggestion
  seq_len = img_wp + txt_wp
```